### DISCUSSION TITLE
Loop over all nodal values of material property in a user object

### DISCUSSION THREAD

**Discussion by FHilty** [idaholab/moose#22602] [Q&A General] [answered]
Hi All,

I am working with discrete nucleation for recrystallization and would like to prevent new grains from overlapping with existing recrystallized grains. My thought to do this was related to the nucleation probability, which is already known to `DiscreteNucleationInserter` and in my model it becomes 0 in recrystallized grains. So I am coding an optional loop that would check if any points inside a proposed nucleus have a 0 probability (or less than some threshold), and if so it would either eliminate the nucleus or resize its radius to not overlap, depending on user options.

What i am unsure of is how to access the value of `_probability` at different elements? I only ever see material property values being looped over the qp n one element. As a follow up question, if I can access any elements material property value is there a way to create a subset of elements to check. Ideally, I would only check elements that fall within a box around the proposed nucleus, and not the entire domain.

Thanks,
Floyd

**Comment by GiudGiud** [comment, marked as answer]
Hello

There have been a few posts on grain nucleation overlap on the discussion forum. Did you have the chance to take a look?

You could block restrict the object instead of looking at the entire domain. 

Guillaume

**Reply by FHilty** [reply]
Hi @GiudGiud after looking into possible solutions what I settled on was using the `RadialAverage` user object. I will create a material property that is 1 in the recrystallized grains and 0 outside, then the `RadialAverage` user object will take that and perform a spatial averaging, such that the value is between 0 and 1 as you approach the new grain areas. In this way when a new grain is being proposed it can check that averaging map to know how close it is to an existing recrystallized grain and adjust accordingly.

I will have to modify the `DiscreteNucleationInserter` to access `RadialAverage` and then it should be straightforward to perform the checks and operations i need.

**Reply by GiudGiud** [reply]
This seems like a working solution. I dont see a spatialValue() API in that user object unfortunately, that would have been the cannon way of accessing the data. 

The issue I think is that this UO only computes a single average, near desired coordinates, instead of the average at every location. Am I correct?

**Reply by FHilty** [reply]
What I found is there is a RadialAverageAux kernel that calls RadialAverage for each point in the domain and stores the result in an Aux variable. So what is collected is a local average for every point in the domain. One issue is the RadialAverageAux is it is registered to the MooseTestApp, so thats an easy enough change for me to use it, but i opened an issue to have it registered to MooseApp so that others can find it and make use the RadialAverage functionality.

**Reply by GiudGiud** [reply]
The NearestPointAverage UO got merged btw, which does have the spatialValue() API which you can call from your object to get local averages
